cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,407 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A structure for encapsulating entire circuits in an operation.
+
+A CircuitOperation is an Operation object that wraps a FrozenCircuit. When
+applied as part of a larger circuit, a CircuitOperation will execute all
+component operations in order, including any nested CircuitOperations.
+"""
+
+from typing import TYPE_CHECKING, AbstractSet, Callable, Dict, List, Optional, Tuple, Union
+
+import dataclasses
+import numpy as np
+
+from cirq import devices, linalg, ops, protocols, study
+from cirq._compat import proper_repr
+from cirq.circuits.frozen_circuit import FrozenCircuit
+from cirq.circuits.insert_strategy import InsertStrategy
+
+if TYPE_CHECKING:
+    import cirq
+
+
+INT_TYPE = Union[int, np.integer]
+
+
+@dataclasses.dataclass(frozen=True, init=False)
+class CircuitOperation(ops.Operation):
+    """An operation that encapsulates a circuit.
+
+    This class captures modifications to the contained circuit, such as tags
+    and loops, to support more condensed serialization. Similar to
+    GateOperation, this type is immutable.
+    """
+
+    circuit: 'cirq.FrozenCircuit'
+    repetitions: int
+    qubit_map: Dict['cirq.Qid', 'cirq.Qid']
+    measurement_key_map: Dict[str, str]
+    param_resolver: study.ParamResolver
+
+    def __init__(
+        self,
+        *contents: 'cirq.OP_TREE',
```

Comment:
That was the original version - I pushed for the OP_TREE because:

1.) OP_TREE contains Circuit, FrozenCircuit as well  
2.) when manually creating CircuitOperation - you'll have to add a bunch of boilerplate unnecessarily

Reply:
Discussed out-of-band: `OP_TREE` support and `to_op` will be reverted.

Reply:
This change has been applied.